File: helper/darwin/KeeWebHelper/InputParser.m
#import <Foundation/Foundation.h>
#import <readline/readline.h>

#import "InputParser.h"
#import "InputCommands/CopyPasteCommand.h"
#import "InputCommands/NoOpCommand.h"
#import "InputCommands/SendKeyCommand.h"
#import "InputCommands/SendTextCommand.h"
#import "InputCommands/UnknownCommand.h"
#import "InputCommands/WaitCommand.h"

@implementation InputParser

- (id<InputCommandBase>)next {
    const char *chars = readline(NULL);
    if (chars) {
        NSString *line = [NSString stringWithUTF8String:chars];
        NSString *command = line;
        NSRange range = [line rangeOfString:@" "];
        if (range.location != NSNotFound) {
            command = [line substringToIndex:range.location];
            line = [line substringFromIndex:range.location + 1];
        }
        if (!command.length) {
            return [[NoOpCommand alloc] init];
        }
        if ([command isEqualToString:@"exit"]) {
            return nil;
        } else if ([command isEqualToString:@"key"]) {
            return [self parseSendCommand:line text:false];
        } else if ([command isEqualToString:@"text"]) {
            return [self parseSendCommand:line text:true];
        } else if ([command isEqualToString:@"copypaste"]) {
            if (line.length) {
                return [[CopyPasteCommand alloc] initWithText:line];
            } else {
                return [[NoOpCommand alloc] init];
            }
        } else if ([command isEqualToString:@"wait"]) {
            NSInteger time = [line integerValue];
            if (time > 0) {
                return [[WaitCommand alloc] initWithTime:time];
            } else {
                return [[NoOpCommand alloc] initWithMessage:@"Bad time"];
            }
        }
        return [[UnknownCommand alloc] initWithName:command];
    } else {
        return nil;
    }
}

- (id<InputCommandBase>)parseSendCommand:(NSString*)args text:(bool)text {
    if (!args.length) {
        return [[NoOpCommand alloc] init];
    }
    CGEventFlags modifier = 0;
    NSMutableCharacterSet *stopChars = [[NSMutableCharacterSet alloc] init];
    [stopChars formUnionWithCharacterSet:[NSCharacterSet decimalDigitCharacterSet]];
    [stopChars formUnionWithCharacterSet:[NSCharacterSet whitespaceCharacterSet]];
    NSInteger index = 0;
    while (index < args.length) {
        unichar firstChar = [args characterAtIndex:index];
        if ([stopChars characterIsMember:firstChar]) {
            break;
        }
        switch (firstChar) {
            case '^':
                modifier |= kCGEventFlagMaskControl;
                break;
            case '+':
                modifier |= kCGEventFlagMaskShift;
                break;
            case '%':
                modifier |= kCGEventFlagMaskAlternate;
                break;
            case '@':
                modifier |= kCGEventFlagMaskCommand;
                break;
            default:
                return [[NoOpCommand alloc] initWithMessage:@"Bad key modifier"];
        }
        index++;
    }
    if (text) {
        index++;
    }
    if (index > 0) {
        if (index >= args.length) {
            return [[NoOpCommand alloc] initWithMessage:@"No key code"];
        }
        args = [args substringFromIndex:index];
    }
    if (text) {
        return [[SendTextCommand alloc] initWithText:args andModifier:modifier];
    } else {
        NSInteger key = [args integerValue];
        if (key <= 0) {
            return [[NoOpCommand alloc] initWithMessage:@"Bad key code"];
        }
        return [[SendKeyCommand alloc] initWithKey:(CGKeyCode)key andModifier:modifier];
    }
}

@end

File: helper/darwin/KeeWebHelper/InputParser.swift
import Foundation

class InputParser {
    class func next() -> InputCommandBase? {
        let maybeStr = readLine()
        if let str = maybeStr {
            let parts = str.characters.split(" ", maxSplit: 1)
            if parts.count == 0 {
                return NoOpCommand()
            }
            let cmd = String(parts[0])
            if cmd.characters.count == 0 {
                return NoOpCommand()
            }
            let cmdArg: String? = parts.count > 1 ? String(parts[1]) : nil
            switch cmd {
            case "exit":
                return nil
            case "key":
                return cmdArg != nil ? parseSendKeyCommand(cmdArg!) : NoOpCommand()
            case "text":
                return cmdArg != nil ? SendTextCommand(text: cmdArg!) : NoOpCommand()
            case "copypaste":
                return cmdArg != nil ? CopyPasteCommand(text: String(parts[1])) : NoOpCommand()
            case "wait":
                let interval = cmdArg != nil ? UInt32(cmdArg!) : nil
                return interval != nil ? WaitCommand(interval: interval!) : NoOpCommand()
            default:
                return UnknownCommand(name: cmd)
            }
        }
        return nil
    }
    
    class func parseSendKeyCommand(text: String) -> InputCommandBase? {
        if text.characters.count == 0 {
            return NoOpCommand()
        }
        var modifiers = ModifierKeys()
        var keyText = text
        while keyText.characters.count > 0 &&
            (keyText.characters.first! < "0" || keyText.characters.first! > "9") {
            let firstChar = keyText.characters.first!
            switch firstChar {
            case "^":
                modifiers.insert(ModifierKeys.ctrl)
            case "+":
                modifiers.insert(ModifierKeys.shift)
            case "%":
                modifiers.insert(ModifierKeys.alt)
            case "@":
                modifiers.insert(ModifierKeys.cmd)
            default:
                fputs("Bad modifier: \(firstChar)\n", stderr)
                return NoOpCommand()
            }
            keyText = keyText.substringFromIndex(keyText.startIndex.advancedBy(1))
        }
        let keyCode = UInt16(keyText)
        if keyCode == nil {
            fputs("Bad key code: \(keyCode)\n", stderr)
            return NoOpCommand()
        }
        return SendKeyCommand(key: keyCode!, modifiers: modifiers)
    }
}
